### PR TITLE
refactor: Move run_transaction params into fields of a TxInput struct and make doc comments of its fields

### PR DESCRIPTION
**Pull request by ltfschoen**
## Purpose of PR

Extension to PR https://github.com/paritytech/parity-ethereum/pull/10742 to address review comment https://github.com/paritytech/parity-ethereum/pull/10742#discussion_r293681775.

## Usage

Check that tests still pass:
```
cargo test -p evmbin -- --nocapture
```

View docs and check that user able to click "info" module and then choose the "TxInput" struct that should be generated in target/doc/parity_evm/info/struct.TxInput.html
```
cargo doc -p evmbin --no-deps --document-private-items --open
```

## Changes made

* [x] Move `run_transaction` parameters into fields of a new `TxInput` struct since comments about parameters of the function do not appear in the documentation, whereas doc comments applied to fields of a public struct are.

Note: Issues that were encountered:
<details><p>
* **NOTE** There is a lot of duplication in declaring instances of the `TxInput` struct in main.rs. I haven't figured out how to refactor it to introduce more code reuse, since only the `informant` field changes.
Note: Already tried making it mutable with let mut tx_input = TxInput { ... }` and then just modifying the value of that field with say `tx_input.informant = display::std_json::Informant::err_only()` before applying that value as an argument of `run_transaction`. 

Example: The following is a snippet of existing code in main.rs
```
// Execute the given transaction and verify resulting state root
// for CLI option `--std-dump-json` or `--std-json`.
if args.flag_std_dump_json || args.flag_std_json {
	if args.flag_std_err_only {
		let tx_input = TxInput {
			name: &name,
			idx,
			spec: &spec,
			pre_state: &pre,
			post_root,
			env_info: &env_info,
			transaction,
			informant: display::std_json::Informant::err_only(),
			trie_spec: &trie_spec,
		};
		// Use Standard JSON informant with err only
		info::run_transaction(tx_input)
	} else if args.flag_std_out_only {
		let tx_input = TxInput {
			name: &name,
			idx,
			spec: &spec,
			pre_state: &pre,
			post_root,
			env_info: &env_info,
			transaction,
			informant: display::std_json::Informant::out_only(),
			trie_spec: &trie_spec,
		};
		// Use Standard JSON informant with out only
		info::run_transaction(tx_input)
	} else {
```

I tried changing it to the following

```
let mut tx_input = TxInput {
	name: &name,
	idx,
	spec: &spec,
	pre_state: &pre,
	post_root,
	env_info: &env_info,
	transaction,
	informant: display::simple::Informant::default(),
	trie_spec: &trie_spec,
};

// Execute the given transaction and verify resulting state root
// for CLI option `--std-dump-json` or `--std-json`.
if args.flag_std_dump_json || args.flag_std_json {
	if args.flag_std_err_only {
		tx_input.informant = display::std_json::Informant::err_only();
		// Use Standard JSON informant with err only
		info::run_transaction(tx_input)
	} else if args.flag_std_out_only {
		tx_input.informant = display::std_json::Informant::out_only();
		// Use Standard JSON informant with out only
		info::run_transaction(tx_input)
	} else {
```

But I get errors 

```
error[E0308]: mismatched types
   --> evmbin/./src/main.rs:221:28
    |
221 |                         tx_input.informant = display::std_json::Informant::err_only();
    |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `display::simple::Informant`, found struct `display::std_json::Informant`
    |
    = note: expected type `display::simple::Informant`
               found type `display::std_json::Informant<std::io::Stderr, std::io::Stderr>`

error[E0308]: mismatched types
   --> evmbin/./src/main.rs:225:28
    |
225 |                         tx_input.informant = display::std_json::Informant::out_only();
    |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `display::simple::Informant`, found struct `display::std_json::Informant`
    |
    = note: expected type `display::simple::Informant`
               found type `display::std_json::Informant<std::io::Stdout, std::io::Stdout>`

error: aborting due to 2 previous errors
```
</p></details>

### CODE REVIEW

File: evmbin/src/info.rs [hunk 1]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
 	})
 }
 
+/// Input data to run transaction
+#[derive(Debug)]
+pub struct TxInput<'a, T> {
+	/// Chain specification name associated with the transaction.
+	pub name: &'a String,
```
Comment (tomusdrw) [inline evmbin/src/info.rs]:
```suggestion
	pub name: &'a str,
```

`&'a str` is less restrictive and will most likely suffice here.

File: evmbin/src/info.rs [hunk 2]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
...
+	pub spec: &'a ethjson::spec::ForkSpec,
+	/// State of all accounts in the system that is a binary tree mapping of each account address to account data
+	/// that is expressed as Plain Old Data containing the account balance, account nonce, account code in bytes,
+	/// and the account storage binary tree map.
+	pub pre_state: &'a pod_state::PodState,
+	/// State root hash associated with the transaction.
+	pub post_root: H256,
+	/// Client environment information associated with the transaction's chain specification.
+	pub env_info: &'a client::EnvInfo,
+	pub transaction: transaction::SignedTransaction,
```
Comment (tomusdrw) [inline evmbin/src/info.rs]: Missing docs?

File: evmbin/src/info.rs [hunk 3]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
...
+	/// JSON formatting informant.
+	pub informant: T,
+	pub trie_spec: &'a TrieSpec,
```
Comment (tomusdrw) [inline evmbin/src/info.rs]: Same

File: evmbin/src/info.rs [hunk 4]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
...
-	pre_state: &pod_state::PodState,
-	// State root hash associated with the transaction.
-	post_root: H256,
-	// Client environment information associated with the transaction's chain specification.
-	env_info: &client::EnvInfo,
-	transaction: transaction::SignedTransaction,
-	// JSON formatting informant.
-	mut informant: T,
-	trie_spec: TrieSpec,
+	tx_input: TxInput<T>
```
Comment (tomusdrw) [inline evmbin/src/info.rs]:
```suggestion
	mut tx_input: TxInput<T>
```

File: evmbin/src/info.rs [hunk 5]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
...
 ) {
-	let spec_name = format!("{:?}", spec).to_lowercase();
-	let spec = match EvmTestClient::spec_from_json(spec) {
+	let mut tx_input = tx_input;
```
Comment (tomusdrw) [inline evmbin/src/info.rs]:
```suggestion
```
This line is redundant, cause you can mark it `mut` in the signature.

Comment (tomusdrw) [reply evmbin/src/info.rs]: The cool thing you can also do to avoid rewriting the rest of the code is to destructure `tx_input`:
```
let TxInput { idx, spec, pre_state, post_root, informant, ... } = tx_input;
```

File: evmbin/src/info.rs [hunk 6]
```diff
@@ -168,14 +175,15 @@ fn dump_state(state: &state::State<state_db::StateDB>) -> Option<pod_state::PodS
 /// Execute EVM with given `ActionParams`.
 pub fn run<'a, F, X>(
 	spec: &'a spec::Spec,
-	trie_spec: TrieSpec,
+	trie_spec: &'a TrieSpec,
```
Comment (tomusdrw) [inline evmbin/src/info.rs]: Is this change strictly required for something? It seems you have to clone the spec anyway, so maybe better take it by value than by reference.

File: evmbin/src/info.rs [hunk 7]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
 	})
 }
 
+/// Input data to run transaction
+#[derive(Debug)]
+pub struct TxInput<'a, T> {
+	/// Chain specification name associated with the transaction.
+	pub name: &'a String,
+	/// Transaction index from list of transactions within a state root hashes corresponding to a chain.
+	pub idx: usize,
```
Comment (dvdplm) [inline evmbin/src/info.rs]: Do we really need all the members to be `pub`? It is generally avoided so that the code is easier to change in the future. Maybe it could be a way to use the `Default` trait to avoid supplying all args all the time? It would also let you do fancy things like have `MyChainTxInput` types where e.g. the `name` member is pre-set and you can have logic to ensure users don't misuse the api (say, use a `ForkSpec` that is invalid for the given chain spec name).

File: evmbin/src/info.rs [hunk 8]
```diff
@@ -100,51 +100,58 @@ pub fn run_action<T: Informant>(
...
-	transaction: transaction::SignedTransaction,
-	// JSON formatting informant.
-	mut informant: T,
-	trie_spec: TrieSpec,
+	tx_input: TxInput<T>
 ) {
-	let spec_name = format!("{:?}", spec).to_lowercase();
-	let spec = match EvmTestClient::spec_from_json(spec) {
+	let mut tx_input = tx_input;
+	let spec_name = format!("{:?}", tx_input.spec).to_lowercase();
```
Comment (dvdplm) [inline evmbin/src/info.rs]: Should `tx_input.spec` perhaps be named `fork_spec` (and `spec_name` be `fork_spec_name`)?

File: evmbin/src/info.rs (hunk 6 above)
Comment (ltfschoen) [reply evmbin/src/info.rs]: I've tried what you recommended but I haven't been able to implement it successfully as it opened a can of worms. This is where I got to https://github.com/paritytech/parity-ethereum/commit/ed5e6674fb08a9b34d61fd4ac996ec2cc7b1c062

<details><p>
   Compiling evmbin v0.1.0 (/home/ltfschoen/code/github/paritytech/parity-ethereum-master/evmbin)
error[E0504]: cannot move `params` into closure because it is borrowed
  --> evmbin/./src/info.rs:95:34
   |
94 |     run(spec, trie_spec, &params.gas, spec.genesis_state(), |mut client| {
   |                           ---------- borrow of `params.gas` occurs here
95 |         let result = match client.call(params, &mut trace::NoopTracer, &mut informant) {
   |                                        ^^^^^^ move into closure occurs here

error[E0504]: cannot move `tx_input` into closure because it is borrowed
   --> evmbin/./src/info.rs:153:43
    |
152 |     let result = run(&spec_checked, trie_spec, &tx_input.transaction.gas, &pre_state, |mut client| {
    |                                                 -------------------- borrow of `tx_input.transaction` occurs here
153 |         let result = client.transact(&env_info, tx_input.transaction, trace::NoopTracer, tx_input.informant);
    |                                                 ^^^^^^^^ move into closure occurs here

error[E0382]: capture of partially moved value: `tx_input`
   --> evmbin/./src/info.rs:153:43
    |
136 |     let TxInput { name, idx, spec, pre_state, post_root, env_info, trie_spec, .. } = tx_input;
    |                                                                    --------- value moved here
...
153 |         let result = client.transact(&env_info, tx_input.transaction, trace::NoopTracer, tx_input.informant);
    |                                                 ^^^^^^^^ value captured here after move
    |
    = note: move occurs because `tx_input.trie_spec` has type `ethcore::TrieSpec`, which does not implement the `Copy` trait

error: aborting due to 3 previous errors
</p></details>

Comment (tomusdrw) [reply evmbin/src/info.rs]: yeah, if you destructure `tx_input` you can't really move it around. just destructure `transaction` as well and pass it?

File: evmbin/src/info.rs (hunk 8 above)
Comment (ltfschoen) [reply evmbin/src/info.rs]: Yes great idea. I definitely think `fork_spec` is a better name than just `spec`. But I think we should rename `spec_name` to `state_test_name` since it refers to one or more state tests that are provided in the JSON file (i.e. add11, create2callPrecompiles)

Comment (ltfschoen) [reply evmbin/src/info.rs]: I've pushed commits with those changes in commits 2e17bb3fde250279d82a1b568ed31c7ecd77312c  and b828e3343049f0e4022e8c11b929d2ae02e66aaa

File: evmbin/src/info.rs (hunk 6 above)
Comment (ltfschoen) [reply evmbin/src/info.rs]: Thanks, that worked, very slick! I've pushed changes in this commit 4b984745c33cd1751e71f0445e8210b0ffa7dedb

File: evmbin/src/info.rs (hunk 7 above)
Comment (ltfschoen) [reply evmbin/src/info.rs]: I didn't figure out how to use the `Default` trait, but I've pushed an approach here  
in this commit https://github.com/paritytech/parity-ethereum/commit/426a4388b0b7989ff2903fe1a07c98ee86642563, but unfortunately I can't work out how to overcome these errors 
<details><p>
error: expected expression, found `}`
   --> evmbin/./src/info.rs:139:3
    |
139 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:146:3
    |
146 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:153:3
    |
153 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:160:3
    |
160 |         }
    |         ^ expected expression

error: expected expression, found `}`
   --> evmbin/./src/info.rs:167:3
    |
167 |         }
    |         ^ expected expression

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:221:7
    |
221 |                         tx_input::std_json_err_only();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:225:7
    |
225 |                         tx_input::std_json_out_only();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:229:7
    |
229 |                         tx_input::std_json_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:237:7
    |
237 |                         tx_input::json_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0433]: failed to resolve: use of undeclared type or module `tx_input`
   --> evmbin/./src/main.rs:241:7
    |
241 |                         tx_input::simple_default();
    |                         ^^^^^^^^ use of undeclared type or module `tx_input`

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:137:15
    |
137 |             informant: display::std_json::Informant::err_only(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stderr, std::io::Stderr>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:136:3
    |
136 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:144:15
    |
144 |             informant: display::std_json::Informant::out_only(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stdout, std::io::Stdout>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:143:3
    |
143 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:151:15
    |
151 |             informant: display::std_json::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::std_json::Informant`
    |
    = note: expected type `T`
               found type `display::std_json::Informant<std::io::Stderr, std::io::Stdout>`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:150:3
    |
150 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:158:15
    |
158 |             informant: display::json::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::json::Informant`
    |
    = note: expected type `T`
               found type `display::json::Informant`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:157:3
    |
157 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error[E0308]: mismatched types
   --> evmbin/./src/info.rs:165:15
    |
165 |             informant: display::simple::Informant::default(),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter, found struct `display::simple::Informant`
    |
    = note: expected type `T`
               found type `display::simple::Informant`

error[E0063]: missing fields `env_info`, `fork_spec_name`, `post_root` and 5 other fields in initializer of `info::TxInput<'_, _>`
   --> evmbin/./src/info.rs:164:3
    |
164 |         TxInput {
    |         ^^^^^^^ missing `env_info`, `fork_spec_name`, `post_root` and 5 other fields

error: aborting due to 20 previous errors
</p></details>
